File: 1353/maximum-number-of-events-that-can-be-attended.java
//根據lee215解釋的解法，主要就一個迴圈，照著 start 日放入event，再去參加「先到期」的活動
class Solution {
    public int maxEvents(int[][] events) {
        Arrays.sort(events, (a,b)->a[0]-b[0]);
//for(int[] event : events){ //印出排序的結果，確認有沒有正確排好
//System.out.println(event[0] + " " + event[1]);
//}
        int ans = 0;
        PriorityQueue<Integer> queue = new PriorityQueue<>( (a,b)->a-b );
        for(int d=1, i=0; d<=100000; d++) {
            while(i<events.length && d>=events[i][0]) { //有跨起開始的時間，用while確認都送入queue
//System.out.println("day:"+d+" event:" + events[i][0] + " " + events[i][1]);
                queue.offer(events[i++][1]); //就加入一筆，裡是結束的時間
            }
//System.out.println("peek() " + queue.peek());

            while(queue.size()>0 && queue.peek()<d) { //過期的活動，逐一丟掉
                queue.poll();
            }
            if(queue.size()<=0) continue; //queue沒有活動可參加，跳過

            if(queue.peek()>=d){ //將到期的還在合理範圍，順利參加一場活動
                queue.poll();
                ans++;
//System.out.println("ans: " + ans);
            }
        }
        return ans;
    }
}
//case 40/44: [[27,27],[8,10],[9,11],[20,21],[25,29],[17,20],[12,12],[12,12],[10,14],[7,7],[6,10],[7,7],[4,8],[30,31],[23,25],[4,6],[17,17],[13,14],[6,9],[13,14]]
//要調一下程式的順序，讓過期的活動先丟掉，以免塞住

File: 1353/maximum-number-of-events-that-can-be-attended.py
# LeetCode 1353. Maximum Number of Events That Can Be Attended
# events 有活動開始、結束的日子，每天只能參加1個活動，問你最多能參加幾個活動
# 處理原則很簡單：(1)活動依序「開始」，放入heap (2)每天參加活動時，選「快結束」的那個
class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        events.sort()  # 先針對活動「開始時間」排序
        ans = 0  # 能順利參加「幾個活動」
        i = 0  # 現在處理的事件是 event[i]
        heap = []  # 某天能參加的活動，都即時在heap裡「可依活動結束時間」取出來
        for day in range(1,100001):  # 活動的日子範圍 1...100000
            while i < len(events) and events[i][0] <= day:  # 活動在 day 之前開始
                heappush(heap, events[i][1])  # 記下活動的結束時間
                i += 1  # 將看下一個活動
            # 現在「所有能參加的活動」都放在heap裡，並標示它的結束時間
            while heap and heap[0] < day:  # 若heap裡「活動已結束」
                heappop(heap)  # 就丟掉
            # 今天要挑什麼活動呢?
            if heap:  # 今天還有活動可以參加，太好了
                heappop(heap)  # 參加最早結束的活動、用掉它
                ans += 1  # 多參加1個活動、真開心
        return ans  # 照此規則「能參加最多的活動」
